### PR TITLE
Resources tab additional references - behind flag

### PR DESCRIPTION
**Pull request by epeach**
Resources Tab: the sequel.
This PR adds map references (1) and references links (1+) to the resources tab. Levelbuilders can already add these to levels, but they will only be displayed behind the flag "additionalResources" at the moment.

Resources tab with a video, map reference, and additional reference link

### CODE REVIEW

File: apps/src/StudioApp.js [hunk 1]
```diff
@@ -1853,7 +1853,7 @@ StudioApp.prototype.configureDom = function (config) {
   var referenceArea = document.getElementById('reference_area');
   // noInstructionsWhenCollapsed is used in TopInstructions to determine when to use CSPTopInstructions (in which case
   // display videos in the top instructions) or CSFTopInstructions (in which case the videos are appended here).
-  const referenceAreaInTopInstructions = config.noInstructionsWhenCollapsed && (experiments.isEnabled('resources_tab') || experiments.isEnabled('resourcesTab'));
+  const referenceAreaInTopInstructions = config.noInstructionsWhenCollapsed && (experiments.isEnabled('resources_tab') || experiments.isEnabled('resourcesTab') || experiments.isEnabled('additionalResources'));
```
Comment (islemaster) [inline apps/src/StudioApp.js:1849]: Obviously somewhat external to this PR, but I count five places we check
```js
experiments.isEnabled('resources_tab') || experiments.isEnabled('resourcesTab')
```
It's probably worth extracting a helper for this - `experiments.isResourcesTabEnabled()` would be fine.  If you're fine with resolving a merge/rebase, we could even do that in a separate PR before merging this change.

File: apps/src/code-studio/appOptions.js [hunk 2]
```diff
@@ -114,6 +114,8 @@
  * @property {boolean} submittable
  * @property {boolean} final_level
  * @property {array} levelVideos
+ * @property {string} mapReference
+ * @property {array} referenceLinks
```
Comment (islemaster) [inline apps/src/code-studio/appOptions.js:118]: :clap: Thanks for keeping the documentation updated!

File: apps/src/StudioApp.js (hunk 1 above)
Comment (epeach) [reply apps/src/StudioApp.js:1849]: In parallel to this PR, I'm working on another one that actually removes all the resourcesTab experiments checks (because we are releasing the resourcesTab to the world). Depending on the review process for both PRs, that one might go out before this one, but probably +/- 1 week of this one, so I don't think adding the helper is much use.

File: apps/src/templates/BaseDialog.jsx [hunk 3]
```diff
@@ -78,6 +79,11 @@ var BaseDialog = React.createClass({
         marginLeft: '-45%'
       });
     }
+    if (this.props.fullHeight) {
+      bodyStyle = Object.assign({}, bodyStyle, {
+        height: '75%',
+      });
```
Comment (islemaster) [inline apps/src/templates/BaseDialog.jsx]: Same behavior with ES6 spread syntax:
```js
bodyStyle = {...bodyStyle, height: '75%'};
```
Although in this case I don't think there's a downside to just mutating the style object:
```js
bodyStyle.height = '75%';
```

File: apps/src/templates/instructions/HelpTabContents.jsx [hunk 4]
```diff
@@ -27,6 +33,22 @@ export default class HelpTabContents extends Component {
             video={this.props.videoData}
           />
         }
+        {experiments.isEnabled('additionalResources') &&
+          this.props.mapReference &&
```
Comment (islemaster) [inline apps/src/templates/instructions/HelpTabContents.jsx]: I like what you did with extracting the `displayReferenceLinks` logic out to the top of the render method.  Could do the same thing here.

File: apps/src/templates/instructions/HelpTabContents.jsx [hunk 5]
```diff
@@ -27,6 +33,22 @@ export default class HelpTabContents extends Component {
...
+          <ResourceLink
+            map={true}
+            reference={this.props.mapReference}
+          />
+        }
+        {displayReferenceLinks &&
+          this.props.referenceLinks.map((link) => (
+            <ResourceLink
+              key={link}
```
Comment (islemaster) [inline apps/src/templates/instructions/HelpTabContents.jsx]: If the links themselves aren't guaranteed to be unique, this could get you in trouble in the edge case where `referenceLinks` contains duplicates.  Maybe safer to use an index:
```jsx
this.props.referenceLinks.map((link, index) => (
  <ResourceLink
    key={index}
    // ...etc...
  />
))
```

File: apps/src/templates/instructions/HelpTabContents.jsx [hunk 6]
```diff
@@ -27,6 +33,22 @@ export default class HelpTabContents extends Component {
             video={this.props.videoData}
           />
         }
+        {experiments.isEnabled('additionalResources') &&
+          this.props.mapReference &&
+          <ResourceLink
+            map={true}
+            reference={this.props.mapReference}
+          />
```
Comment (islemaster) [inline apps/src/templates/instructions/HelpTabContents.jsx]: If you make the `map` boolean prop in `ResourceLink` not required, you can treat it like a flag here and below:
```jsx
// `map` is passed to `ResourceLink` as true.
<ResourceLink map reference={this.props.mapReference}/>
// `map` is passed to `ResourceLink` as undefined.
<ResourceLink key={key} reference={link}/>
```

File: apps/src/templates/instructions/ResourceLink.jsx [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+  };
+
+  getLinkText = (data) => {
+    this.setState({text: data});
+  };
+
+  componentWillMount = () => {
+    // Gets the text content from the meta description tag from the
+    // reference to which we're linking
+    $.get(this.props.reference, (data) => { this.getLinkText($(data).filter("meta[name='description']").attr("content"));});
```
Comment (islemaster) [inline apps/src/templates/instructions/ResourceLink.jsx]: This is super-cool, kind of confusing, should probably not be on one line and I don't see the test coverage for it.  Can you walk me through this?

File: apps/src/templates/instructions/ResourceLink.jsx [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+    const { map } = this.props;
+
+    const iconStyle = map ? {...styles.commonIcon, ...styles.mapIcon} : {...styles.commonIcon, ...styles.resourceIcon};
+    const thumbnailStyle = map ? {...styles.commonThumbnail, ...styles.mapThumbnail} : {...styles.commonThumbnail};
+
+    return (
+      <div>
+        <div style={styles.resourceStyle} onClick={this.selectResource}>
+          <span style={thumbnailStyle}>
+              <i style={iconStyle} className={map ? "fa fa-map" : "fa fa-book"}/>
```
Comment (islemaster) [inline apps/src/templates/instructions/ResourceLink.jsx]: See also [our `FontAwesome` component](https://github.com/code-dot-org/code-dot-org/blob/staging/apps/src/templates/FontAwesome.jsx).

File: dashboard/app/models/levels/level.rb [hunk 9]
```diff
@@ -477,6 +477,14 @@ def summary_for_lesson_plans
     summary
   end
 
+  def additional_reference_links
+    links = {
+      mapReference: map_reference,
+      referenceLinks: reference_links
+    }
+    links
```
Comment (islemaster) [inline dashboard/app/models/levels/level.rb]: `links` is unneeded, I believe - can be expressed as
```rb
def additional_reference_links
  {
    mapReference: map_reference,
    referenceLinks: reference_links
  }
end
```
On a related note, why is this needed? Seems like this could be easily handled in levels_helper.rb where a mapping is occurring anyway.

File: dashboard/app/helpers/levels_helper.rb [hunk 10]
```diff
@@ -236,6 +236,8 @@ def app_options
     # Sets video options for this level
     if @app_options[:level]
       @app_options[:level][:levelVideos] = @level.related_videos.map(&:summarize)
+      @app_options[:level][:mapReference] = @level.additional_reference_links[:mapReference]
+      @app_options[:level][:referenceLinks] = @level.additional_reference_links[:referenceLinks]
```
Comment (islemaster) [inline dashboard/app/helpers/levels_helper.rb]: Comment on line 236 no longer sufficiently describes this if-block.

**Comment by epeach** [comment]
@islemaster Ready for another look when you get a chance - this also have the NetworkResourceLink in here which is in this approved [PR](https://github.com/code-dot-org/code-dot-org/pull/21246). Once that PR finished circle and goes into staging, I'll merge staging back in here to remove those files from the 'changed' list. If you'd like to wait to review until then, that's cool.
